src/Microsoft.AspNetCore.Blazor.Browser.JS/src/Interop/DotNetInvoker.ts

```diff
@@ -0,0 +1,103 @@
+﻿import { MethodOptions, DotnetMethodArgumentsList } from './MethodOptions';
+import { platform } from '../Environment';
+import { System_String } from '../Platform/Platform';
+import { getRegisteredFunction } from './RegisteredFunction';
+
+export function invokeDotNetMethod<T>(methodOptions: MethodOptions, ...args: any[]): (T | null) {
+    const method = platform.findMethod(
+        "Microsoft.AspNetCore.Blazor.Browser",
+        "Microsoft.AspNetCore.Blazor.Browser.Interop",
+        "JavaScriptInvoke",
+        "InvokeDotnetMethod");
+
+    const packedArgs = packArguments(args);
+
+    const serializedOptions = platform.toDotNetString(JSON.stringify(methodOptions));
+    const serializedArgs = platform.toDotNetString(JSON.stringify(packedArgs));
+    const serializedResult = platform.callMethod(method, null, [serializedOptions, serializedArgs]);
+
+    if (serializedResult !== null && serializedResult !== undefined && (serializedResult as any) !== 0) {
+        const result = JSON.parse(platform.toJavaScriptString(serializedResult as System_String));
+        return result;
+    }
+
+    return null;
+}
+
+function packArguments(args: any[]): DotnetMethodArgumentsList {
+    var result = {};
+    if (args.length == 0) {
+        return result;
+    }
+
+    if (args.length > 7) {
+        for (let i = 0; i < 7; i++) {
+            result[`argument${[i + 1]}`] = args[i];
+        }
+        result['argument8'] = packArguments(args.slice(7));
+    } else {
+        for (let i = 0; i < args.length; i++) {
+            result[`argument${[i + 1]}`] = args[i];
+        }
+    }
+
+    return result;
+}
+
+let globalId = 0;
+
+export function invokeDotNetMethodAsync<T>(methodOptions: MethodOptions, ...args: any[]): Promise<T | null> {
+    const resolveId = (globalId++).toString();
+    const rejectId = (globalId++).toString();
+    methodOptions.async = { resolveId: resolveId, rejectId: rejectId, functionName: "invokeJavaScriptCallback" };
+
+    const result = new Promise<T | null>((resolve, reject) => {
+        TrackedReference.track(resolveId, resolve);
+        TrackedReference.track(rejectId, reject);
+    });
+
+    invokeDotNetMethod(methodOptions, ...args);
+
+    return result;
+}
+
+export function invokeJavaScriptCallback(id: string, ...args: any[]): void {
+    const callbackRef = TrackedReference.get(id);
+    const callback = callbackRef.trackedObject as Function;
+    callback.apply(null, args);
+}
+
+class TrackedReference {
+    private static references: Map<string, any> = new Map<string, any>();
+
+    private constructor(public id: string, public trackedObject: any) {
```

Comment:
It looks like we never read back the `.id` property value from any `TrackedReference` instance. If that is the case, maybe we could avoid constructing `TrackedReference` instances altogether.

Instead of having `TrackedReference` as an instantiatable class, we could only have the `references` map, and it could directly map from `id` to the `trackedObject` without any intermediate wrapper around the `trackedObject`. It would save one JS object allocation per tracked object.

Also it's interesting to see the `Map` class in use here. Elsewhere I've been using raw objects typed as `[{key: string}: any]`, which has the benefit of not requiring ES6, but has the cost of requiring `hasOwnProperty` checks around property accesses. Since [browser support for `Map` looks good back to IE11](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map#Browser_compatibility), I think this is a good improvement and I'll try to use `Map` elsewhere from now on.

Reply:
@SteveSandersonMS I've used index signatures in the past, but given current browser support its fair to just use ES6 IMO and ask that you polyfill if you have to.